### PR TITLE
Pod templates as their own type

### PR DESCRIPTION
**Pull request by smarterclayton**
This pull request adds pod templates as their own type (first step towards #170). Having
a distinct type allows their use to be clearly delineated for end users, to have their
own independent validation rules, and to simplify interactions from the client. This does
not prevent pods from having "inert" behavior in the future that can also be used for
templates.  Replication controllers should work against inert pods or pod templates equally.

Only the last commit is new - the others are issues found while trying to support a
type that only exists in v1beta3.

### CODE REVIEW

**Comment by pmorie** [comment]
@liggitt since we were just talking about this

**Comment by bgrant0607** [comment]
Thanks!!

The main change is now the next-to-last commit. :-)

I'm fine with the API. I can see the benefit of an explicit template.

One thing to consider for separate templates: We may want to allow users to specify the resourceVersion of the template in the ObjectReference (the necessary field already exists) and force the replication controller to observe the specified version before it creates any new pods, for WAW consistency, if the user cares.

When this is ready for full review, let me know, though I may need to punt it to someone else.

**Comment by bgrant0607** [comment]
I'm fine with this being v1beta3 only. I want to finalize v1beta3 and flip to it as the default asap.

Looking at the json, I see why you changed "spec" to "template". My main concern is whether we might add status in the future, such as to publish the last time the template was used to create a pod.

**Comment by smarterclayton** [comment]
I could certainly imagine a status in the future.  Is an object that behaves like a template having a root "template" field better  (obviousness) vs encouraging others to invent things besides spec. My thought was that spec is a real object which is converged, while a template isn't.  Might be good to get a few other opinions.

**Comment by bgrant0607** [comment]
@smarterclayton You're logic is reasonable. I don't really want to bikeshed on this particular topic. Let's go with template.

**Comment by piosz** [comment]
Is this PR active? If not please close it according to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/pull-requests.md

**Comment by smarterclayton** [comment]
Yes

**Comment by smarterclayton** [comment]
Rebased and ready to use.

**Comment by smarterclayton** [comment]
Ready for review

**Comment by bgrant0607** [comment]
Cool, thanks. Will take a look.

You're getting a gendocs failure due to the addition of a new noun.

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -888,6 +888,24 @@ func ValidatePodStatusUpdate(newPod, oldPod *api.Pod) errs.ValidationErrorList {
 	return allErrs
 }
 
+// ValidatePodTemplate tests if required fields in the pod template are set.
+func ValidatePodTemplate(pod *api.PodTemplate) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	allErrs = append(allErrs, ValidateObjectMeta(&pod.ObjectMeta, true, ValidatePodName).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidatePodTemplateSpec(&pod.Template, 0).Prefix("spec")...)
```
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: This field is called "template" now.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -888,6 +888,24 @@ func ValidatePodStatusUpdate(newPod, oldPod *api.Pod) errs.ValidationErrorList {
...
+	return allErrs
+}
+
+// ValidatePodTemplateUpdate tests to see if the update is legal for an end user to make. newPod is updated with fields
+// that cannot be changed.
+func ValidatePodTemplateUpdate(newPod, oldPod *api.PodTemplate) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	allErrs = append(allErrs, ValidateObjectMetaUpdate(&oldPod.ObjectMeta, &newPod.ObjectMeta).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidatePodTemplateSpec(&newPod.Template, 0).Prefix("spec")...)
```
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: This field is called "template" now.

**Comment by bgrant0607** [comment]
Looks pretty good. Found only one minor problem.

The next steps are to expose a subresource for creating pods from the template, and then to use that in the replication controller?

**Comment by smarterclayton** [comment]
Yes

----- Original Message -----

File: pkg/registry/podtemplate/etcd/etcd.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+)
+
+// rest implements a RESTStorage for pod templates against etcd
+type REST struct {
+	etcdgeneric.Etcd
+}
+
+// NewREST returns a RESTStorage object that will work against pod templates.
+func NewREST(h tools.EtcdHelper) *REST {
```
Comment (bgrant0607) [inline pkg/registry/podtemplate/etcd/etcd.go]: Love how simple this is now.

**Comment by smarterclayton** [comment]
Comments addressed.

File: pkg/registry/podtemplate/etcd/etcd.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/registry/podtemplate/etcd/etcd.go]: Yeah, reduced about 5k lines from openshift code base.

**Comment by bgrant0607** [comment]
Please regenerate the kubectl docs.

**Comment by bgrant0607** [comment]
LGTM. Waiting for Travis.

**Comment by bgrant0607** [comment]
```
# github.com/GoogleCloudPlatform/kubernetes/pkg/registry/controller/etcd
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/registry/controller/etcd/etcd_test.go:78: unknown api.PodTemplate field 'Spec' in struct literal
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/registry/controller/etcd [build failed]
```

**Comment by smarterclayton** [comment]
Weird.

**Comment by bgrant0607** [comment]
Rerunning Travis. I have no idea what happened. Looks like Template to me.

**Comment by bgrant0607** [comment]
Hmm. Still fails.

**Comment by bgrant0607** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/blob/1a2e4f8e1d35b38fde55414a3c61347f5894a24f/pkg/registry/controller/etcd/etcd_test.go#L63

**Comment by smarterclayton** [comment]
Fixed, merge changes coming in.

----- Original Message -----

**Comment by bgrant0607** [comment]
Thanks. Tests passed, but needs to be rebased, apparently.

**Comment by smarterclayton** [comment]
Rebased and travis is green

**Pull request merged**